### ISSUE TITLE
Line numbers in code on tutorials

### ISSUE THREAD

**Issue by vec4** [chunhualiao/freeCompilerCamp#117]
It would be nice to have line numbers on code blocks for tutorials, as sometimes it is helpful to refer to the code in the tutorial text.

**Comment by chunhualiao** [comment]
Which tutorials are you referring to?  Some do already have the line
numbers.

**Comment by vec4** [comment]
Yes, you're right - I hadn't noticed that some do have them (e.g., https://github.com/freeCompilerCamp/freecompilercamp.github.io/blob/master/_posts/2019-08-29-rose-ast-traversal.markdown). It seems that they are hard-coded there, which is fine, but it should be possible to have them automatically generated through Rouge.

**Comment by chunhualiao** [comment]
Can you share one example using Rouge?  We can always try new things like
this.  The criteria is to improve the overall productivity of us.

**Comment by vec4** [comment]
I think we are already using Rouge - it is the default highlighter for Jekyll 3. Line numbers can be added easily by using `{% highlight c++ linenos %} insert code here... {% endhighlight %}` and then doing some CSS styling to make it look nice/how we want. See: https://aimuke.github.io/guides/2019-06-01-line-numbers/ as an example.

I tried adding this to one of the tutorials locally, without any additional styling, and it works but looks ugly. I can play around with the CSS in my free time. The only downside is that we would have to use the liquid `{% highlight %}` tags instead of the three ticks ```. But we can always just switch to the liquid tags whenever we think that line numbers would be useful for a code snippet and just continue using the three ticks otherwise.

**Comment by vec4** [comment]
Support has been added for line numbers in code. See here for an example: http://freecompilercamp.org/rose-complex-types-practical/. The syntax is a little more complicated, but we can continue to use the three backticks for whenever we do not need line numbers. Here's an example of the syntax:

```
<figure class="lineno-container">
{% highlight <language> linenos %}
<your_code_here>
{% endhighlight %}
</figure>
```

The `lineno-container` class is needed in order to style it differently from clickable command boxes.

I updated the contribute page to include some information on this as part of the commit.

As a side note, if we want to print line numbers with the `cat` command, we can use the `-n` flag; e.g., `cat -n`. This is useful when printing some code in the terminal, as it makes it more readable.
